### PR TITLE
ssp: SSCR reset - hardware specific flow

### PR DESCRIPTION
**Pull request by ArsenEloglian**
hardware specific flow using bit 20 for reset

Signed-off-by: Arsen Eloglian <ArsenX.Eloglian@intel.com>

### CODE REVIEW

**Review by marcinszkudlinski — DISMISSED** [review]

**Review by lgirdwood — DISMISSED** [review]

File: drivers/dai/intel/ssp/ssp.c [hunk 1]
```diff
@@ -706,7 +706,7 @@ static inline void dai_ssp_pm_runtime_en_ssp_clk_gating(struct dai_intel_ssp *dp
 
 static void dai_ssp_pm_runtime_en_ssp_power(struct dai_intel_ssp *dp, uint32_t index)
 {
-#if CONFIG_SOC_SERIES_INTEL_CAVS_V25
+#if CONFIG_SOC_SERIES_INTEL_CAVS_V25 || !CONFIG_INTEL_ADSP_CAVS
```
Comment (andyross) [inline drivers/dai/intel/ssp/ssp.c:709]: FWIW: this kind of platform test is generally best expressed with things like a devicetree property or a CONFIG_HAS_* kconfig, etc...

File: drivers/dai/intel/ssp/ssp.c [hunk 2]
```diff
@@ -1538,7 +1538,11 @@ static int dai_ssp_set_config_blob(struct dai_intel_ssp *dp, const void *spec_co
 	ssrsa = blob->i2s_driver_config.i2s_config.ssrsa;
 
 	sys_write32(ssc0, dai_base(dp) + SSCR0);
+	sys_write32(blob->i2s_driver_config.i2s_config.ssc2 & (~(BIT(20))),
+			dai_base(dp) + SSCR2); /* hardware specific flow */
 	sys_write32(blob->i2s_driver_config.i2s_config.ssc1, dai_base(dp) + SSCR1);
+	sys_write32(blob->i2s_driver_config.i2s_config.ssc2 | (BIT(20)),
+			dai_base(dp) + SSCR2); /* hardware specific flow */
```
Comment (andyross) [inline drivers/dai/intel/ssp/ssp.c:1545]: Wouldn't hurt to put that BIT(20) into the header somewhere and give it a symbolic name.  From context it looks like some kind of dis/enable bit to protect the write to SSCR1?

File: drivers/dai/intel/ssp/ssp.c [hunk 3]
```diff
@@ -1748,12 +1752,16 @@ static void dai_ssp_stop(struct dai_intel_ssp *dp, int direction)
 	/* disable SSP port if no users */
 	if (ssp->state[DAI_DIR_CAPTURE] == DAI_STATE_PRE_RUNNING &&
 	    ssp->state[DAI_DIR_PLAYBACK] == DAI_STATE_PRE_RUNNING) {
+#if CONFIG_INTEL_ADSP_CAVS
 		if (!(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)) {
+#endif /* CONFIG_INTEL_ADSP_CAVS */
```
Comment (andyross) [inline drivers/dai/intel/ssp/ssp.c:1757]: Not loving this syntax, putting the if() and brackets into the conditional means that the indentation level now becomes context-dependent.

How about:

```c
if (!IS_ENABLED(CONFIG_INTEL_ADSP_CAVS) || !(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)) {
    // ...
}
```
Alternatively you could define a `bool clear_rse_bits;` or something to gate the tests, and set that based on preprocessor state.

**Review by andyross — COMMENTED** [review]
A few nitpicks.  The only one gating my +1 is the indentation lie on !CAVS.

File: drivers/dai/intel/ssp/ssp.c [hunk 4]
```diff
@@ -1748,7 +1752,12 @@ static void dai_ssp_stop(struct dai_intel_ssp *dp, int direction)
 	/* disable SSP port if no users */
 	if (ssp->state[DAI_DIR_CAPTURE] == DAI_STATE_PRE_RUNNING &&
 	    ssp->state[DAI_DIR_PLAYBACK] == DAI_STATE_PRE_RUNNING) {
-		if (!(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)) {
+#if CONFIG_INTEL_ADSP_CAVS
+	bool clear_TRSE_RSRE = true;
+#else
+	bool clear_TRSE_RSRE = false;
+#endif /* CONFIG_INTEL_ADSP_CAVS */
+		if (clear_TRSE_RSRE && !(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)) {
```
Comment (marcinszkudlinski) [inline drivers/dai/intel/ssp/ssp.c:1760]: if (clear_rse_bits)

File: drivers/dai/intel/ssp/ssp.c [hunk 5]
```diff
@@ -1748,7 +1752,12 @@ static void dai_ssp_stop(struct dai_intel_ssp *dp, int direction)
 	/* disable SSP port if no users */
 	if (ssp->state[DAI_DIR_CAPTURE] == DAI_STATE_PRE_RUNNING &&
 	    ssp->state[DAI_DIR_PLAYBACK] == DAI_STATE_PRE_RUNNING) {
-		if (!(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)) {
+#if CONFIG_INTEL_ADSP_CAVS
+	bool clear_TRSE_RSRE = true;
+#else
+	bool clear_TRSE_RSRE = false;
```
Comment (marcinszkudlinski) [inline drivers/dai/intel/ssp/ssp.c:1758]: clear_rse_bits

File: drivers/dai/intel/ssp/ssp.h [hunk 6]
```diff
@@ -130,6 +130,8 @@
 #define SSCR2_LJDFD		BIT(17)
 #define SSCR2_MMRATF		BIT(18)
 #define SSCR2_SMTATF		BIT(19)
+define SSCR2_SFRMEN		BIT(20)
```
Comment (marcinszkudlinski) [inline drivers/dai/intel/ssp/ssp.h:133]: #define

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]

**Review by marcinszkudlinski — DISMISSED** [review]

File: drivers/dai/intel/ssp/ssp.c (hunk 3 above)
Comment (ArsenEloglian) [reply drivers/dai/intel/ssp/ssp.c:1759]: > 

For me, it's more readable to use if/endif statement, but using a macro is also a good idea...

Comment (ArsenEloglian) [reply drivers/dai/intel/ssp/ssp.c]: > Or this:
> 
> ```
> bool clear_rse_bits = COND_CODE_1(CONFIG_INTEL_ADSP_CAVS, (!(ssp->clk_active & SSP_CLK_BCLK_ES_REQ)), (false));
> ```

Done :)

**Review by marcinszkudlinski — DISMISSED** [review]

**Review by andyross — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Review by abonislawski — DISMISSED** [review]

**Comment by kv2019i** [comment]
@ArsenEloglian I don't think this should be closed. Rather we carry this on a local tree until upstream release freeze is over and only then this can proceed. But I think this PR can be kept open. @ceolin @nashif can you confirm?

**Comment by ceolin** [comment]
> @ArsenEloglian I don't think this should be closed. Rather we carry this on a local tree until upstream release freeze is over and only then this can proceed. But I think this PR can be kept open. @ceolin @nashif can you confirm?

yes, please keep it open. We probably will need a rebase / re-approval but the changes were already agreed.

**Review by ceolin — DISMISSED** [review]

**Review by marcinszkudlinski — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**